### DISCUSSION TITLE
How to add force to a model?

### DISCUSSION THREAD

**Discussion by zhang-qiang-github** [InteractiveComputerGraphics/PositionBasedDynamics#155] [Q&A]
I am learning the `pyPBD/examples/beam_model.py`, and I have several questions:

**Question1** How to add a cylinder?
**Question2** The one end of the beam is fixed, and the other end of the beam is sway. Why the other end is sway? Is it affected by the gravity? Where set the gravity in the code?
**Question3** How to apply a force to the beam? Please have a look for the following figure:

![image](https://user-images.githubusercontent.com/29271051/229275204-3b13fb45-0ab7-4df9-9c44-9c59f59d2812.png)

I expect to apply a force to a tube (the red arrow), and the tube turn to be bending. How can I apply a force to the tube?

Any suggestion is appreciated~~~

**Comment by janbender** [comment]
Q1: Generate a tet mesh of a cylinder (e.g. using tools like Tetgen) and load the vertices and tets using the command "addTetModel".
Q2: Yes, gravity is acting on the beam. I just recognized that the Python binding for setting gravity is missing. I will add that in the next version.
Q3: You can get the particle data from the SimulationModel by "getParticles()". In the particle data you can set the acceleration for each particle which would be force/mass.

**Reply by zhang-qiang-github** [reply]
Thanks for kindly reply. 
Q1: If I want to change the beam to a cylinder, is the tetrahedral necessary? Or triangle also can be used?
Q2: I have compiled the C++ source code, and I find the is defined as `m_gravitation` in `Simulation` class. But, there is no public method to get/set gravity. Is there any method can change the gravity in C++?
Q3: How to change the acceleration? Is there any example/demo?

**Comment by janbender** [comment]
Q1: if you use a triangle mesh, the object will be hollow. For volumetric objects you need a volumetric representation.

Q2: This is a generic parameter. You can get the value by

sim->getVecValue<Real>(Simulation::GRAVITATION)

and set it analogously.

Q3: First get the particle data from the model. Then call setAcceleration(particleIndex, accel).

**Reply by zhang-qiang-github** [reply]
> Q3: How to apply a force to the beam?

For the force, I expect the beam turn to be bending when the force is applied. When the force is removed, the beam would turn to be straight.

But by `setAcceleration`, the beam would be bending firstly, then it would automatically restore straight. 

I hope the added force would be balanced with the resistance of beam, so that the beam would keep the bending situation. The beam would be static until another force is applied.

Is it possible to achive the above purpose?
